### PR TITLE
Use relative links for commits, mentions, and issues in markdown

### LINKED ISSUE #29404: markdown: wrong hostname in commit url with empty ROOT_URL
### Description

when i leave ROOT_URL empty
then gitea uses `http://localhost` as hostname in commit urls
but gitea should produce urls without hostname

ping @KN4CK3R via #28909

repro https://try.gitea.io/milahu/test-gitea-markdown-wrong-commit-url

part of #19345

### Gitea Version

1.21.6

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

gitea runs in a container on localhost, accessed from localhost
so gitea thinks that its hostname is "localhost", but that is wrong

### Database

None

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #29404

Use relative links for
- commits
- mentions
- issues

### CODE REVIEW

**Comment by lunny** [comment]
No, except you can know all the renderers will not be used on mail templates.

**Review by delvh — APPROVED** [review]
The markdown renderer is only run for the UI and not for the API, right?

But yeah, I completely misunderstood the original request, I thought it talked about these headers (`Reviewed-on`, …) in commit messages.

**Comment by KN4CK3R** [comment]
@lunny is right, mails would be broken. Fixed that and added a small test to prevent it in future.

File: services/mailer/mail_test.go [hunk 1]
```diff
@@ -96,6 +104,12 @@ func TestComposeIssueCommentMessage(t *testing.T) {
 	assert.Equal(t, "<user2/repo1/issues/1/comment/2@localhost>", gomailMsg.GetHeader("Message-ID")[0], "Message-ID header doesn't match")
 	assert.Equal(t, "<mailto:"+replyTo+">", gomailMsg.GetHeader("List-Post")[0])
 	assert.Len(t, gomailMsg.GetHeader("List-Unsubscribe"), 2) // url + mailto
+
+	var buf strings.Builder
+	gomailMsg.WriteTo(&buf)
+	mailMsg := strings.ReplaceAll(strings.ReplaceAll(buf.String(), "=", ""), "\r\n", "") // naïve quoted printable decoding, to allow string searching
```
Comment (milahu) [inline services/mailer/mail_test.go]: nit: use ascii "naive"

Comment (silverwind) [reply services/mailer/mail_test.go]: Either way is fine, but https://english.stackexchange.com/a/12748/418296 calls it "if not actually archaic" 😆.

**Review by silverwind — APPROVED** [review]

File: services/mailer/mail_test.go (hunk 1 above)
Comment (KN4CK3R) [reply services/mailer/mail_test.go]: They are talking about US context. When I studied the `naïve` was everywhere. 😏

**Review by lunny — APPROVED** [review]

File: services/mailer/mail_test.go (hunk 1 above)
Comment (lunny) [reply services/mailer/mail_test.go]: Which language is that? French or German?

**Comment by KN4CK3R** [comment]
Don't know why the unit tests don't run in the workflow but running them locally revealed some errors in the tests. Reason is this:
https://github.com/go-gitea/gitea/blob/e55926ebfe88d6ee079842967dc7dccc2a9cdbf2/modules/markup/markdown/markdown_test.go#L24-L26
That's not how our `setting.AppURL` and `setting.AppSubURL` work. Will fix this later.

**Comment by lunny** [comment]
I encountered a similar problem in #29222, the test data are inconsistent.

**Comment by KN4CK3R** [comment]
Added #29459 to fix the tests.

File: services/mailer/mail_test.go (hunk 1 above)
Comment (silverwind) [reply services/mailer/mail_test.go]: Let's keep it, golang files support unicode.

Comment (KN4CK3R) [reply services/mailer/mail_test.go]: Noticed I can just pass the whole mail source into the `quotedprintable.Reader` and not just the content part.

File: services/mailer/mail.go [hunk 2]
```diff
@@ -222,7 +222,8 @@ func composeIssueCommentMessages(ctx *mailCommentContext, lang string, recipient
 	body, err := markdown.RenderString(&markup.RenderContext{
 		Ctx: ctx,
 		Links: markup.Links{
-			Base: ctx.Issue.Repo.HTMLURL(),
+			AbsolutePrefix: true,
```
Comment (KN4CK3R) [inline services/mailer/mail.go:225]: The mail, maybe `/routers/common/markup.go` and/or rendered webhook messages should be the only places with `AbsolutePrefix = true`. Some tests have `true` too but that should be changed in another PR.

**Pull request merged**